### PR TITLE
Vectorize - step 2 max spanning tree fix

### PR DESCRIPTION
**Pull request by csarofeen**
Vectorize branch without this change
image

This branch:
image

There's still room for improvement on vectorization. The remaining issue is if the reference has an inner most dimension that's broadcasted that's resolved by other tensors, we cannot vectorize correctly.

This PR allows the vectorization pass to compute the max spanning tree because of: 
https://github.com/csarofeen/pytorch/pull/2124/files#diff-cb1f12c937ff579f33d244faf2152ecf9a4a28a8ca7203863bcf158de149f8eeR731

We don't want to prioritize consumer->producer->consumer paths where the producer->consumer resolves a broadcast.

This code is mostly refactoring to avoid replication of logic. We strip all the logic effectively out of the propagate paths and put them in the max spanning compute paths. We disable recording when we're in compute paths, but we reenable it during propagate paths.

This is because compute paths will dictate max spanning tree, then when we run the spanning tree we actually want to compute the vectorization factors.

### CODE REVIEW

File: benchmarks/cpp/nvfuser/batch_norm_channels_first_backward.cpp [hunk 1]
```diff
@@ -310,18 +310,19 @@ NVFUSER_BENCHMARK_DEFINE(
...
+    // ->Args({128, 128, 56})
+    // ->Args({128, 256, 28})
+    // ->Args({128, 512, 28})
     ->Args({128, 512, 14})
-    ->Args({128, 1024, 14})
-    ->Args({128, 1024, 7})
-    ->Args({128, 2048, 7})
+    // ->Args({128, 1024, 14})
+    // ->Args({128, 1024, 7})
+    // ->Args({128, 2048, 7})
```
Comment (csarofeen) [inline benchmarks/cpp/nvfuser/batch_norm_channels_first_backward.cpp]: revert

File: torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp [hunk 2]
```diff
@@ -230,22 +240,206 @@ ContiguousInnerDimensionsMapper::ContiguousInnerDimensionsMapper(
...
+    TensorView* reference,
+    const std::vector<IterDomain*>& reference_ids,
+    std::shared_ptr<const ComputeAtMap> ca_map,
+    const std::unordered_set<Split*>& divisible_splits) {
+
+  return ContiguousInnerDimensionsMapper(
+      reference, reference_ids, ca_map, divisible_splits);
+}
+
+void ContiguousInnerDimensionsMapper::propagateExtentSplitBackward(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:269]: Just pulled this code into their own functions. Shouldn't be anything but code movement.

File: torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h [hunk 3]
```diff
@@ -385,46 +385,77 @@ class TORCH_CUDA_CU_API ProjectedExtent {
...
+//   MaxInfoSpanningTree::Propagator::propagateC2P
+//
+// The challenge here is the information we need for
+// MaxInfoSpanningTree::computeInfoC2P is the same information we need to
+// compute for MaxInfoSpanningTree::Propagator::propagateC2P
+//
+// We could compute both of these passes at the same time, only saving the
+// result produced from processing the edge that's chosen from
+// MaxInfoSpanningTree::computeInfoC2P while processing based on
+// MaxInfoSpanningTree::Propagator::propagateC2P. Hover, this would require
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h]: nit: You meant this?
```suggestion
// MaxInfoSpanningTree::Propagator::propagateC2P. However, this would require
```

File: torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp [hunk 4]
```diff
@@ -881,11 +944,42 @@ void ContiguousInnerDimensionsMapper::propagateSibling(
...
+}
+
+// MaxInfoSpanningTree functions
+void ContiguousInnerDimensionsMapper::propagateC2P(
+    TensorView* from,
+    TensorView* to) {
+  recording_ = true;
+  auto from_info = tv_infos_.at(from);
+  auto to_info = computeInfoC2P(from, to, from_info);
+  tv_infos_[to] = to_info;
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:963]: Do you need to set `recording_` back to false when you are done with `computeInfoC2P`?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:963]: Is there a chance the max spanning tree would be recomputed? I thought max spanning tree is computed once only calling `computeInfo...` calls, then once that's done we'll never be computing the spanning tree again so we can just leave recording on.

File: torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h [hunk 5]
```diff
@@ -438,24 +469,96 @@ class TORCH_CUDA_CU_API ContiguousInnerDimensionsMapper
...
+    }
 
-  std::unordered_map<TensorView*, std::vector<IterDomain*>>::iterator
-  projectIdToRFactor(TensorView* ref, std::vector<IterDomain*> ids);
+    bool operator<(const Information& other_info) const final {
+      auto other_mapped_domain = dynamic_cast<const MappedDomain&>(other_info);
+      return mapped_root_ids_.size() >
+          other_mapped_domain.mapped_root_ids_.size() &&
+          mapped_rfactor_ids_.size() >
+          other_mapped_domain.mapped_rfactor_ids_.size();
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h]: Should this be `<` instead of `>`?  In `MaxRootDomainInfoSpanningTree::Information`, we are using `<`.
https://github.com/csarofeen/pytorch/blob/2057f37541e0c3e898094d6cb8246756520853ac/torch/csrc/jit/codegen/cuda/maxinfo_propagator.cpp#L164-L176

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h]: I'm sure you're right, but for some reason it doesn't seem to make any difference whatsoever if it's `>` or `<`. In fact I also tried just plain `true` and `false` and no matter what I pick here it seems to work as required. I can't seem to bottom out on this why, but if you want to play around with it, this is a test where you can see the mappings aren't great on the `vectorize` branch (the other vectorize PR), but works fine on this branch.

```.cpp
// Test propogation doesn't proceed across missing dimensions
TEST_F(NVFuserTest, FusionVectorizeSpanningTree_CUDA) {
  auto fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  auto bcast_inp = makeContigConcreteTensor({2});
  fusion.addInput(bcast_inp);
  auto bcast = broadcast(bcast_inp, {false, true});

  std::vector<TensorView*> intermediates;
  std::vector<TensorView*> outputs;
  for (auto i : c10::irange(10)) {
    auto resolution_inp = makeContigConcreteTensor({2, 2});
    fusion.addInput(resolution_inp);
    auto intermediate = add(bcast, resolution_inp);
    if (i > 0) {
      auto output = add(intermediates.back(), intermediate);
      outputs.push_back(output);
      fusion.addOutput(output);
    }
    intermediates.push_back(intermediate);
  }
    
  fusion.printMath();

  auto mapper = vectorize_helper::ContiguousInnerDimensionsMapper::map(
      outputs[0], {outputs[0]->axis(0), outputs[0]->axis(1)});

  for (auto tv : ir_utils::allTvs(&fusion)) {
    std::cout << tv->toString() << " :: ";
    for (auto axis : tv->getRootDomain()) {
      std::cout << mapper.getMappedExtent(axis).toString() << ", ";
    }
    std::cout << std::endl;
  }
}
```

It seems no matter what, we never propagate from the `bcast` `TensorView` in to anything but the input with dim 1 in this problem. Before this PR, the mappings printed from this test is:
```
T0_g[ iS0{2} ] :: {} / {}, 
T1_l[ iS1{2}, bS2{1} ] :: {} / {}, {} / {}, 
T2_g[ iS3{2}, iS4{2} ] :: {2} / {}, {2} / {}, 
T3_l[ iS5{2}, iS6{2} ] :: {2} / {}, {2} / {}, 
T4_g[ iS7{2}, iS8{2} ] :: {2} / {}, {2} / {}, 
T5_l[ iS9{2}, iS10{2} ] :: {2} / {}, {2} / {}, 
T6_g[ iS11{2}, iS12{2} ] :: {2} / {}, {2} / {}, 
T7_g[ iS13{2}, iS14{2} ] :: {} / {}, {} / {}, 
T8_l[ iS15{2}, iS16{2} ] :: {} / {}, {} / {}, 
T9_g[ iS17{2}, iS18{2} ] :: {2} / {}, {2} / {}, 
T10_g[ iS19{2}, iS20{2} ] :: {} / {}, {} / {}, 
T11_l[ iS21{2}, iS22{2} ] :: {} / {}, {} / {}, 
T12_g[ iS23{2}, iS24{2} ] :: {} / {}, {} / {}, 
T13_g[ iS25{2}, iS26{2} ] :: {} / {}, {} / {}, 
T14_l[ iS27{2}, iS28{2} ] :: {} / {}, {} / {}, 
T15_g[ iS29{2}, iS30{2} ] :: {} / {}, {} / {}, 
T16_g[ iS31{2}, iS32{2} ] :: {} / {}, {} / {}, 
T17_l[ iS33{2}, iS34{2} ] :: {} / {}, {} / {}, 
T18_g[ iS35{2}, iS36{2} ] :: {} / {}, {} / {}, 
T19_g[ iS37{2}, iS38{2} ] :: {} / {}, {} / {}, 
T20_l[ iS39{2}, iS40{2} ] :: {} / {}, {} / {}, 
T21_g[ iS41{2}, iS42{2} ] :: {} / {}, {} / {}, 
T22_g[ iS43{2}, iS44{2} ] :: {} / {}, {} / {}, 
T23_l[ iS45{2}, iS46{2} ] :: {} / {}, {} / {}, 
T24_g[ iS47{2}, iS48{2} ] :: {} / {}, {} / {}, 
T25_g[ iS49{2}, iS50{2} ] :: {} / {}, {} / {}, 
T26_l[ iS51{2}, iS52{2} ] :: {} / {}, {} / {}, 
T27_g[ iS53{2}, iS54{2} ] :: {} / {}, {} / {}, 
T28_g[ iS55{2}, iS56{2} ] :: {} / {}, {} / {}, 
T29_l[ iS57{2}, iS58{2} ] :: {} / {}, {} / {}, 
T30_g[ iS59{2}, iS60{2} ] :: {} / {}, {} / {}, 
```
So you can see a lot of mappings broke because it was propagated through T1, however on this PR, it's correctly:
```.cpp
T0_g[ iS0{2} ] :: {} / {}, 
T1_l[ iS1{2}, bS2{1} ] :: {} / {}, {} / {}, 
T2_g[ iS3{2}, iS4{2} ] :: {2} / {}, {2} / {}, 
T3_l[ iS5{2}, iS6{2} ] :: {2} / {}, {2} / {}, 
T4_g[ iS7{2}, iS8{2} ] :: {2} / {}, {2} / {}, 
T5_l[ iS9{2}, iS10{2} ] :: {2} / {}, {2} / {}, 
T6_g[ iS11{2}, iS12{2} ] :: {2} / {}, {2} / {}, 
T7_g[ iS13{2}, iS14{2} ] :: {2} / {}, {2} / {}, 
T8_l[ iS15{2}, iS16{2} ] :: {2} / {}, {2} / {}, 
T9_g[ iS17{2}, iS18{2} ] :: {2} / {}, {2} / {}, 
T10_g[ iS19{2}, iS20{2} ] :: {2} / {}, {2} / {}, 
T11_l[ iS21{2}, iS22{2} ] :: {2} / {}, {2} / {}, 
T12_g[ iS23{2}, iS24{2} ] :: {2} / {}, {2} / {}, 
T13_g[ iS25{2}, iS26{2} ] :: {2} / {}, {2} / {}, 
T14_l[ iS27{2}, iS28{2} ] :: {2} / {}, {2} / {}, 
T15_g[ iS29{2}, iS30{2} ] :: {2} / {}, {2} / {}, 
T16_g[ iS31{2}, iS32{2} ] :: {2} / {}, {2} / {}, 
T17_l[ iS33{2}, iS34{2} ] :: {2} / {}, {2} / {}, 
T18_g[ iS35{2}, iS36{2} ] :: {2} / {}, {2} / {}, 
T19_g[ iS37{2}, iS38{2} ] :: {2} / {}, {2} / {}, 
T20_l[ iS39{2}, iS40{2} ] :: {2} / {}, {2} / {}, 
T21_g[ iS41{2}, iS42{2} ] :: {2} / {}, {2} / {}, 
T22_g[ iS43{2}, iS44{2} ] :: {2} / {}, {2} / {}, 
T23_l[ iS45{2}, iS46{2} ] :: {2} / {}, {2} / {}, 
T24_g[ iS47{2}, iS48{2} ] :: {2} / {}, {2} / {}, 
T25_g[ iS49{2}, iS50{2} ] :: {2} / {}, {2} / {}, 
T26_l[ iS51{2}, iS52{2} ] :: {2} / {}, {2} / {}, 
T27_g[ iS53{2}, iS54{2} ] :: {2} / {}, {2} / {}, 
T28_g[ iS55{2}, iS56{2} ] :: {2} / {}, {2} / {}, 
T29_l[ iS57{2}, iS58{2} ] :: {2} / {}, {2} / {}, 
T30_g[ iS59{2}, iS60{2} ] :: {2} / {}, {2} / {}, 
```
No matter what choice is made for the information comparator function!!!

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h]: Tried building out a better test that reverses ordering, and what tensor we're starting from for the mapping:
```.cpp

// Test propogation doesn't proceed across missing dimensions
TEST_F(NVFuserTest, FusionVectorizeSpanningTree_CUDA) {
  auto fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  std::vector<TensorView*> inputs;
  std::vector<TensorView*> intermediates;
  std::vector<TensorView*> outputs;

  auto bcast_inp = makeContigConcreteTensor({2});
  inputs.push_back(bcast_inp);
  auto bcast = broadcast(bcast_inp, {false, true});

  for (auto i : c10::irange(10)) {
    auto resolution_inp = makeContigConcreteTensor({2, 2});
    inputs.push_back(resolution_inp);
    auto intermediate = add(bcast, resolution_inp);
    if (i > 0) {
      auto output = add(intermediates.back(), intermediate);
      outputs.push_back(output);
    }
    intermediates.push_back(intermediate);
  }

  for (auto rev_inp : {false, true}) {
    for (auto rev_out : {false, true}) {
      // Clear fusion inputs / outputs
      {
        auto fusion_outs = fusion.outputs();
        for (auto out : fusion_outs) {
          fusion.removeOutput(out);
        }
        auto fusion_inps = fusion.inputs();
        for (auto inp : fusion_inps) {
          fusion.removeOutput(inp);
        }
      }

      if (rev_inp) {
        std::reverse(inputs.begin(), inputs.end());
      }

      if (rev_out) {
        std::reverse(outputs.begin(), outputs.end());
      }

      {
        // Populate outputs and inputs
        for (auto out : outputs) {
          fusion.addOutput(out);
        }

        for (auto inp : inputs) {
          fusion.addInput(inp);
        }
      }

      for (auto out : outputs) {
        auto mapper = vectorize_helper::ContiguousInnerDimensionsMapper::map(
            out, {out->axis(0), out->axis(1)});

        for (auto tv : ir_utils::allTvs(&fusion)) {
          if (tv->name() == 0 || tv->name() == 1) {
            continue;
          }
          for (auto axis : tv->getRootDomain()) {
            TORCH_INTERNAL_ASSERT(
                mapper.getMappedExtent(axis).getNumerator()->evaluateInt() ==
                2);
            TORCH_INTERNAL_ASSERT(
                mapper.getMappedExtent(axis).getDenominator()->evaluateInt() ==
                1);
          }
        }
      }
    }
  }
}
```
